src/html.cr

```diff
@@ -114,47 +114,83 @@ module HTML
   # HTML.unescape("Crystal &amp; You") # => "Crystal & You"
   # ```
   def self.unescape(string : String) : String
-    string.gsub(/&(?:([a-zA-Z0-9]{2,32};?)|\#([0-9]+);?|\#[xX]([0-9A-Fa-f]+);?)/) do |string, match|
-      if code = match[1]?
-        # Try to find the code
-        value = named_entity(code)
-
-        unless value || code.ends_with?(';')
-          # If we can't find it and it doesn't end with ';',
-          # we need to find each prefix of it.
-          # We start from the largest prefix.
-          removed = 0
-          until code.empty?
-            code = code.rchop
-            removed += 1
-
-            value = named_entity(code)
-            if value
-              # If we find it, we need to append the part that
-              # isn't part of the matched code
-              value += string[-removed..-1]
-              break
-            end
-          end
-        end
-
-        # We either found the code or not,
-        # in which case we need to return the original string
-        value || string
-      elsif code = match[2]?
-        # Find by decimal code
-        decode_codepoint(code.to_i) || string
-      elsif code = match[3]?
-        # Find by hexadecimal code
-        decode_codepoint(code.to_i(16)) || string
-      else
-        string
-      end
+    return string unless string.includes?('&')
+
+    String.build(string.bytesize) do |io|
+      unescape(string.to_slice, io)
+    end
+  end
+
+  private def self.unescape(slice, io)
+    while bytesize = slice.index('&'.ord)
+      io.write(slice[0, bytesize])
+      slice += bytesize &+ 1
+
+      ptr = unescape_entity(slice.to_unsafe, io)
+      slice += ptr - slice.to_unsafe
     end
+
+    io.write slice
   end
 
-  private def self.named_entity(code)
-    HTML::SINGLE_CHAR_ENTITIES[code]? || HTML::DOUBLE_CHAR_ENTITIES[code]?
+  private def self.unescape_entity(ptr, io)
+    if '#' === ptr.value
+      unescape_numbered_entity(ptr, io)
+    else
+      unescape_named_entity(ptr, io)
+    end
+  end
+
+  private def self.unescape_numbered_entity(ptr, io)
+    start_ptr = ptr
+
+    ptr += 1
+
+    hex = ptr.value.unsafe_chr.in?('x', 'X')
+    if hex
+      ptr += 1
+      base = 16
+    else
+      base = 10
+    end
+
+    x = 0_u32
+
+    # skip leading zeros
+    while ptr.value === '0'
+      ptr += 1
+    end
+    number_start_ptr = ptr
+
+    while digit = ptr.value.unsafe_chr.to_i?(base)
+      # The number of consumed digits is limited to the representation of
+      # Char::MAX_CODEPOINT which is below that of UInt32::MAX
+      x &*= base
+      x &+= digit
+
+      ptr += 1
+    end
+
+    if ptr - number_start_ptr > 8
+      # size exceeds maxlength, so it can't be a valid codepoint and might have
+      # overflow.
+      x = 0_u32
+    end
+
+    size = ptr - start_ptr
+    unless size > 2 && (char = decode_codepoint(x))
```

Comment:
`size == 2` if the input string is `&#9;`, but it should be transformed to `\t`. I don't think `size == 2` is correct here

Reply:
Ah, I see. Yeah `size > 2` is only correct for hex digits. I'll fix it.